src/file/BlockFS.cpp

```diff
@@ -0,0 +1,279 @@
+    boost::filesystem::path p(location());
+
+    data_array_dir = Directory(p / data_arrays, file->fileMode());
+    tag_dir = Directory(p / tags, file->fileMode());
+    multi_tag_dir = Directory(p / multi_tags, file->fileMode());
+    source_dir = Directory(p / sources, file->fileMode());
+}
+
+//--------------------------------------------------
+// Methods concerning sources
+//--------------------------------------------------
+
+bool BlockFS::hasSource(const std::string &name_or_id) const {
+    return getSource(name_or_id) != nullptr;
+}
+
+
+std::shared_ptr<base::ISource> BlockFS::getSource(const std::string &name_or_id) const {
+    shared_ptr<base::ISource> source;
+    boost::optional<boost::filesystem::path> path = source_dir.findByNameOrAttribute("entity_id", name_or_id);
+    if (path) {
+        return make_shared<SourceFS>(file(), path->string());
+    }
+    return source;
+}
+
+
+ndsize_t BlockFS::sourceCount() const {
+    return source_dir.subdirCount();
+}
+
+
+std::shared_ptr<base::ISource> BlockFS::getSource(ndsize_t index) const {
+    if (index >= sourceCount()) {
+        throw OutOfBounds("Trying to access block.source with invalid index.", index);
+    }
+    boost::filesystem::path p = source_dir.sub_dir_by_index(index);
+    shared_ptr<SourceFS> source;
+    source = make_shared<SourceFS>(file(), p.string());
+    return source;
+}
+
+
+std::shared_ptr<base::ISource> BlockFS::createSource(const std::string &name, const std::string &type) {
+    if (name.empty()) {
+        throw EmptyString("name");
+    }
+    if (hasSource(name)) {
+        throw DuplicateName("createSource");
+    }
+    string id = util::createId();
+    SourceFS s(file(), source_dir.location(), id, type, name);
+    return make_shared<SourceFS>(s);
+}
+
+
+bool BlockFS::deleteSource(const std::string &name_or_id) {
+    return source_dir.removeObjectByNameOrAttribute("entity_id", name_or_id);
+}
+
+//--------------------------------------------------
+// Methods concerning data arrays
+//--------------------------------------------------
+
+bool BlockFS::hasDataArray(const std::string &name_or_id) const {
+    return getDataArray(name_or_id) != nullptr;
+}
+
+
+std::shared_ptr<base::IDataArray> BlockFS::getDataArray(const std::string &name_or_id) const {
+    shared_ptr<base::IDataArray> array;
+    boost::optional<boost::filesystem::path> path = data_array_dir.findByNameOrAttribute("entity_id", name_or_id);
+    if (path) {
+        DataArrayFS a(file(), block(), path->string());
+        return make_shared<DataArrayFS>(a);
+    }
+    return array;
+}
+
+
+std::shared_ptr<base::IDataArray> BlockFS::getDataArray(ndsize_t index) const {
+    if (index >= dataArrayCount()) {
+        throw OutOfBounds("Trying to access block.dataArray with invalid index.", index);
+    }
+    boost::filesystem::path p = data_array_dir.sub_dir_by_index(index);
+    DataArrayFS a(file(), block(), p.string());
+    return make_shared<DataArrayFS>(a);
```

Comment:
`make_shared<DataArrayFS>(file(), block(), p.string())`. Otherwise we use the copy-constructor, and then have to destruct `a` again.